src/main/java/com/flink/connector/BigQuerySourceFunction.java

```diff
@@ -0,0 +1,54 @@
+package com.flink.connector;
+
+import org.apache.flink.api.common.serialization.DeserializationSchema;
+import org.apache.flink.api.common.serialization.RuntimeContextInitializationContextAdapters;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.java.typeutils.ResultTypeQueryable;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.streaming.api.functions.source.RichSourceFunction;
+import org.apache.flink.table.data.RowData;
+
+import com.google.cloud.bigquery.storage.v1.BigQueryReadClient;
+
+
+public final class BigQuerySourceFunction extends RichSourceFunction<RowData> implements ResultTypeQueryable<RowData> {
+	private static final long serialVersionUID = 1;
+
+	private final String project_id;
+	private final String dataset;
+	private final String table;
+	private final int snapshotMillis;
+	private final DeserializationSchema<RowData> deserializer;
+	public BigQuerySourceFunction(String project_id, String dataset, String table, int snapshotMillis,
+			DeserializationSchema<RowData> deserializer) {
+		this.project_id = project_id;
+		this.dataset = dataset;
+		this.table = table;
+		this.snapshotMillis = snapshotMillis;
+		this.deserializer = deserializer;
+	}
+
+	@Override
+	public TypeInformation<RowData> getProducedType() {
+		return deserializer.getProducedType();
+	}
+
+	@Override
+	public void open(Configuration parameters) throws Exception {
+		deserializer.open(RuntimeContextInitializationContextAdapters.deserializationAdapter(getRuntimeContext()));
+	}
+
+	@Override
+	public void run(SourceContext<RowData> ctx) throws Exception {
+		BiqQueryReadSession readSession = new BiqQueryReadSession();
```

Comment:
Doesn't make sense - the read session should be created on the main app (driver/master), readRows should be invoked on the executor